### DISCUSSION TITLE
Wrong switch toggle after a certain time

### DISCUSSION THREAD

**Discussion by LeeWiu** [arendst/Tasmota#15172] [Q&A] [answered]
Hi, I flashed a Gosund P1 with Tasmota and applyed template and rules  https://templates.blakadder.com/gosund_P1.html
It's working fine but sometimes in the morning, when I press physical button 3, will switch off the second relay, then I press again physical button for 2-3 times and working fain again.
How can I debug this behavior?
Maybe it's something wrong with this rule ? 
Rule2 ON Analog#A0div10<30 DO Power3 2 BREAK ON Analog#A0div10<60 DO Power2 2 BREAK ON Analog#A0div10<80 DO Power1 2 ENDON

**Comment by sfromis** [comment]
For debugging of what happens when, follow the console after setting `Weblog 3` (or other logging destinations, also with log level 3).

**Reply by barbudor** [reply]
If you want to fine tune the rule you can proceed as below:

Disable the rule with `Rule2 0`
Go to the main web page where you can see the value of `Analog0`. When no button is pressed, it should display 1024.
Press and keep pressed Button 1, wait a little bit and note the value as A. If the value are slighly changing, try to make a rough average.
Release Button 1, Press and keep pressed Button 2, and do the same => note the value as B
Release Button 2, Press and keep pressed Button 3, and do the same => note the value as C

Calculate :
X = (1024 + A) / 20
Y = (A + B ) / 20
Z = (B + C) / 20
Update the rule with the value of Z, Y, X to replace the values 30, 60 and 80
Re-enable the rule with `Rule2 1`

You may do that once in the morning and another time later. In that case the 1st time do not replace the values yet, just keep X1, Y1, Z1
On the 2nd time calculate X2, Y2, Z2
Then update the rule using the max value for each couple (X = max(X1, X2); ...)

**Reply by LeeWiu** [reply]
Wow, thanks for instructions, I made the first set and values are slightly different:
X = (1024 + A) / 20 | 88.8 vs 80
Y = (A + B ) / 20 | 61,5 vs 60
Z = (B + C) / 20 | 35.15 vs 30

I will try a few more times tomorow.

**Comment by LeeWiu** [comment]
This are the final values: 88.85, 61.6, 35.3
Thank you again!

**Comment by LeeWiu** [comment, marked as answer]
I spoke to soon, after 2 days I get the same behaviour: button3 turned off second relay. I look the value of Analog0 for all buttons and are still the same.
There is another method to turn on/off switches?

**Reply by sfromis** [reply]
As the "Tasmota" name has "M" for "MQTT", that is a primary method for controlling Tasmota devices. 

If you want to work on improving the control via the analog input, you could do something like creating rules to discard changed analog values if the next cycle does not land on the same side of your threshold. The thing is that there may be a short rise/fall time when the button is hit, and if Tasmota samples in that short interval before the new value is settled, that could mean that the wrong action is performed. Effectively implementing something like that Tasmota ButtonDebounce feature for normal digital inputs.

Depending on how the hardware is built, it may be doable to rewire the buttons to connect to standard digital inputs, if you like to have fun doing a small modification...

**Reply by LeeWiu** [reply]
Seams to complicated for me to rewire.
You mean that I can't control the switches from physical buttons if wifi connection is down?

**Reply by sfromis** [reply]
No, that's not what I meant. [Rules](https://tasmota.github.io/docs/Rules/) are a feature local on each Tasmota device, takes a bit of a learning curve to get started.

Completely separate from this, beware that Tasmota may not work well if it can't get real time, typically from a NTP server.

**Reply by barbudor** [reply]
@TeHashX so if button 3 triggers relay 2 it means that the analog value is higher than expected, higher than 35.
What are your last B (button 2) and C (button 3) values? You may increase the 35 to get closer to  B/20 but not too close.

Don't forget also that it does not support pressing 2 buttons at the same. Always be sure to press only one.

**Comment by steffengy** [comment]
Given that even with different values I also saw misfires, I opted to try a debounce:
rule2
```
ON Analog#A0div10<35 DO backlog ruletimer4 0.5; ruletimer5 0; ruletimer6 0 BREAK 
ON Analog#A0div10<61 DO backlog ruletimer4 0; ruletimer5 0.5; ruletimer6 0 BREAK 
ON Analog#A0div10<88 DO backlog ruletimer4 0; ruletimer5 0; ruletimer6 0.5 BREAK 
ON Analog#A0div10>88 DO backlog ruletimer4 0; ruletimer5 0; ruletimer6 0 DONE`
```
rule3
`ON rules#timer=4 DO POWER3 2 BREAK ON rules#timer=5 DO POWER2 2 BREAK ON rules#timer=6 DO POWER1 2 BREAK`
